crates/red_knot_python_semantic/resources/mdtest/call/methods.md

```diff
@@ -255,4 +255,81 @@ method_wrapper()
 method_wrapper(C(), C, "one too many")
 ```
 
+## `@classmethod`
+
+### Basic
+
+When a `@classmethod` attribute is accessed, it returns a bound method object, even when accessed on
+the class object itself:
+
+```py
+from __future__ import annotations
+
+class C:
+    @classmethod
+    def f(cls: type[C], x: int) -> str:
+        return "a"
+
+reveal_type(C.f)  # revealed: <bound method `f` of `Literal[C]`>
+reveal_type(C().f)  # revealed: <bound method `f` of `type[C]`>
+```
+
+The `cls` method argument is then implicitly passed as the first argument when calling the method:
+
+```py
+reveal_type(C.f(1))  # revealed: str
+reveal_type(C().f(1))  # revealed: str
+```
+
+When the class method is called incorrectly, we detect it:
+
+```py
+C.f("incorrect")  # error: [invalid-argument-type]
+C.f()  # error: [missing-argument]
+C.f(1, 2)  # error: [too-many-positional-arguments]
+```
+
+If the `cls` parameter is wrongly annotated, we emit an error at the call site:
+
+```py
+class D:
+    @classmethod
+    def f(cls: D):
```

Comment:
I would expect this `D` name reference to fail to resolve, but somehow it doesn't? This isn't a stub file, and doesn't have `from __future__ import annotations`, so name references in annotations should resolve eagerly, and the name `D` should not be bound here yet.

Reply:
> I would expect this `D` name reference to fail to resolve, but somehow it doesn't?

Ah, I was wondering about this, because one of my other examples failed when I ran it through CPython. I then added `from __future__ import annotations`, but forgot to write it down as a (pre existing?) issue. I'll do so.

Reply:
This particular test already includes `from __future__ import annotations` at the top.

Reported as https://github.com/astral-sh/ruff/issues/16341